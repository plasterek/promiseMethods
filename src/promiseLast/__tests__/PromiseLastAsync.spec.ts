import { promiseLastASync } from "../PromiseLastASync.service";

describe("promiseLastASync function", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  const resolvedPromise: () => Promise<unknown> = () => Promise.resolve("resolved");
  const delayedResolvedPromise: () => Promise<unknown> = () =>
    new Promise((resolve) => {
      setTimeout(function () {
        resolve("resolvedWithDelay");
      }, 3000);
    });
  const rejectedPromise: () => Promise<unknown> = () => Promise.reject("rejected");
  describe("When all Promises in array are resolved", () => {
    it("It returns Promise with result that took most time to resolve", () => {
      //given
      jest.useFakeTimers();
      jest.spyOn(global, "setTimeout");
      const promiseArray: (() => Promise<unknown>)[] = [delayedResolvedPromise, resolvedPromise];
      //when
      const results: Promise<unknown> = promiseLastASync(promiseArray);
      jest.runAllTimers();
      //then
      expect.assertions(1);
      expect(results).resolves.toMatchSnapshot();
    });
  });
  describe("When any promise rejects in given array", () => {
    it("It will return rejected promise after all promises in given array are resolved", () => {
      //given
      jest.useFakeTimers();
      jest.spyOn(global, "setTimeout");
      const promiseArray: (() => Promise<unknown>)[] = [delayedResolvedPromise, rejectedPromise];
      //when
      const results: Promise<unknown> = promiseLastASync(promiseArray);
      jest.runAllTimers();
      //then
      expect.assertions(1);
      expect(results).rejects.toMatchSnapshot();
    });
  });
  describe("When given array is empty", () => {
    it("It returns Promise with an empty array", () => {
      //given
      const promiseArray: any[] = [];
      //when
      const results: Promise<unknown> = promiseLastASync(promiseArray);
      //then
      expect.assertions(1);
      expect(results).resolves.toMatchSnapshot();
    });
  });
});
