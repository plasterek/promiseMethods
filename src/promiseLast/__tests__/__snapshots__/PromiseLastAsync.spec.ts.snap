// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`promiseLastASync function When all Promises in array are resolved It returns Promise with result that took most time to resolve 1`] = `"resolvedWithDelay"`;

exports[`promiseLastASync function When any promise rejects in given array It will return rejected promise after all promises in given array are resolved 1`] = `"rejected"`;

exports[`promiseLastASync function When given array is empty It returns Promise with an empty array 1`] = `[]`;
